In [2]:
from numpy import loadtxt
from xgboost import XGBClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import pandas as pd
from gensim import utils
import gensim.parsing.preprocessing as gsp
from gensim.models.doc2vec import TaggedDocument, Doc2Vec
from sklearn.base import BaseEstimator
from sklearn import utils as skl_utils
from tqdm import tqdm
import re
import multiprocessing
import numpy as np

In [3]:
fake_real_data = pd.read_csv("fake_job_postings.csv")

In [4]:
fake_real_data.head()

job_id                                      title            location  \
0       1                           Marketing Intern    US, NY, New York   
1       2  Customer Service - Cloud Video Production      NZ, , Auckland   
2       3    Commissioning Machinery Assistant (CMA)       US, IA, Wever   
3       4          Account Executive - Washington DC  US, DC, Washington   
4       5                        Bill Review Manager  US, FL, Fort Worth   

  department salary_range                                    company_profile  \
0  Marketing          NaN  We're Food52, and we've created a groundbreaki...   
1    Success          NaN  90 Seconds, the worlds Cloud Video Production ...   
2        NaN          NaN  Valor Services provides Workforce Solutions th...   
3      Sales          NaN  Our passion for improving quality of life thro...   
4        NaN          NaN  SpotSource Solutions LLC is a Global Human Cap...   

                                         description  \
0  Food52, a fast-growing, James Beard Award-winn...   
1  Organised - Focused - Vibrant - Awesome!Do you...   
2  Our client, located in Houston, is actively se...   
3  THE COMPANY: ESRI – Environmental Systems Rese...   
4  JOB TITLE: Itemization Review ManagerLOCATION:...   

                                        requirements  \
0  Experience with content management systems a m...   
1  What we expect from you:Your key responsibilit...   
2  Implement pre-commissioning and commissioning ...   
3  EDUCATION: Bachelor’s or Master’s in GIS, busi...   
4  QUALIFICATIONS:RN license in the State of Texa...   

                                            benefits  telecommuting  \
0                                                NaN              0   
1  What you will get from usThrough being part of...              0   
2                                                NaN              0   
3  Our culture is anything but corporate—we have ...              0   
4                              Full Benefits Offered              0   

   has_company_logo  has_questions employment_type required_experience  \
0                 1              0           Other          Internship   
1                 1              0       Full-time      Not Applicable   
2                 1              0             NaN                 NaN   
3                 1              0       Full-time    Mid-Senior level   
4                 1              1       Full-time    Mid-Senior level   

  required_education                   industry              function  \
0                NaN                        NaN             Marketing   
1                NaN  Marketing and Advertising      Customer Service   
2                NaN                        NaN                   NaN   
3  Bachelor's Degree          Computer Software                 Sales   
4  Bachelor's Degree     Hospital & Health Care  Health Care Provider   

   fraudulent  
0           0  
1           0  
2           0  
3           0  
4           0

In [5]:
fake_real_data.iloc[0,6]

'Food52, a fast-growing, James Beard Award-winning online food community and crowd-sourced and curated recipe hub, is currently interviewing full- and part-time unpaid interns to work in a small team of editors, executives, and developers in its New York City headquarters.Reproducing and/or repackaging existing Food52 content for a number of partner sites, such as Huffington Post, Yahoo, Buzzfeed, and more in their various content management systemsResearching blogs and websites for the Provisions by Food52 Affiliate ProgramAssisting in day-to-day affiliate program support, such as screening affiliates and assisting in any affiliate inquiriesSupporting with PR &amp; Events when neededHelping with office administrative work, such as filing, mailing, and preparing for meetingsWorking with developers to document bugs and suggest improvements to the siteSupporting the marketing and executive staff'

In [6]:
fake_real_data = fake_real_data.dropna()

In [7]:
#Gensim preprocessing filters to remove numberc values, tags, punctuation, multiple whitespaces and stopwords
filters = [
           gsp.strip_tags, 
           gsp.strip_punctuation,
           gsp.strip_multiple_whitespaces,
           gsp.strip_numeric,
           gsp.remove_stopwords, 
           gsp.strip_short, 
           gsp.stem_text
          ]
#Function to clean the description
def clean_text(data):
    data = data.lower()
    data = utils.to_unicode(data)
    for fil in filters:
        data = fil(data)
    return data

In [8]:
#Example run of clean text
clean_text(fake_real_data.iloc[0,6])

'respons manag english speak editori team build team best class editorsset content creation schedul ensur deadlin adher toresearch write latest tech topic new relat android ecosystemensur content site consist high qualityb face voic url adbddeccedeefeceeaa'

In [9]:
#Split the data into x and Y
train_data = fake_real_data['description']
test_data = fake_real_data['fraudulent']
print("len of df_x",len(train_data))
print("len of df_y",len(test_data))


len of df_x 774
len of df_y 774


In [10]:
#Vectorize the attributes using TF IDF format to weight each word in a document
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf_transformer = TfidfVectorizer()
#Clean the text and remove stop words
train_data = train_data.apply(lambda x : clean_text(x))
#Fit the model and transform based of TFIDF vectorization
tfidf_vectors = tfidf_transformer.fit(train_data).transform(train_data)

In [11]:
tfidf_vectors.shape

(774, 6226)

In [12]:
#Use pipeline to classify documents based on the attributes and find the accuracy score with respect to data_y
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score
predict_tf_idf = Pipeline(steps=[('tfidf',tfidf_transformer),
                         ('xgboost', XGBClassifier(objective='binary:logistic'))])
#cross validate the prediction of pipeline with test column
scores = cross_val_score(predict_tf_idf, train_data, test_data, cv=5)
print('Accuracy for Tf-Idf & XGBoost Classifier using description : ', scores.mean())

Accuracy for Tf-Idf & XGBoost Classifier using description :  0.9586761625471303


In [15]:
df = pd.read_csv("fake_job_postings.csv",engine='python',error_bad_lines=False)

In [16]:
df[df['fraudulent']==1].shape

(866, 18)

In [17]:
dat = {'text': []}
data = pd.DataFrame(data=dat)

In [18]:
df.title.fillna('', inplace=True)
df.company_profile.fillna('', inplace=True)
df.description.fillna('', inplace=True)
df.requirements.fillna('', inplace=True)

data['text'] = df['title'] + ' ' + df['company_profile'] + ' ' + df['description'] + ' ' + df['requirements']
data
data['fraudulent'] = df['fraudulent']
data[data['fraudulent']==1].shape

(866, 2)

In [19]:
from nltk.corpus import stopwords
import nltk
from nltk.stem.snowball import SnowballStemmer
from nltk.stem import WordNetLemmatizer
nltk.download('stopwords')
nltk.download('wordnet')
nltk.download('punkt')
stop_words = set(stopwords.words('english')) 
from nltk.tokenize import word_tokenize

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\sdrsh\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping corpora\stopwords.zip.
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\sdrsh\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping corpora\wordnet.zip.
[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\sdrsh\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping tokenizers\punkt.zip.


In [20]:
stemmer = SnowballStemmer("english")
lem=WordNetLemmatizer()
data['text'] = data['text'].map(lambda x: re.sub(r"\W", " ",str(x)))
data['text'] = data['text'].map(lambda x: re.sub(r'[,\.!?]', " ",str(x)))
data['text'] = data['text'].map(lambda x: re.sub(r" \d+", " ",str(x)))
data['text'] = data['text'].map(lambda x: word_tokenize(x.lower()))
data['text'] = data['text'].map(lambda x: ' '.join(x))
data['text'] = data['text'].apply(lambda x: ' '.join([word for word in x.split() if word not in stop_words]))
data['lemm'] = data['text'].apply(lambda x: ' '.join([lem.lemmatize(y) for y in x.split()]))
data['stemm'] = data['lemm'].apply(lambda x: ' '.join([stemmer.stem(y) for y in x.split()]))
data['text'] = data['stemm']

# data['text']=data['text'].str.lower()
# data['text']=data['text'].str.replace('\d+', '')
# data['text']=data['text'].str.replace(r"http\S+", '')
# data
# df[5690]
data

text  fraudulent  \
0      market intern food52 creat groundbreak award w...           0   
1      custom servic cloud video product second world...           0   
2      commiss machineri assist cma valor servic prov...           0   
3      account execut washington dc passion improv qu...           0   
4      bill review manag spotsourc solut llc global h...           0   
5      account clerk job overviewapex environment con...           0   
6      head content f found theâ fonpit agâ rose inte...           0   
7      lead guest servic specialist airenvyâ mission ...           0   
8      hp bsm sme solutions3 woman own small busi who...           0   
9      custom servic associ part time novitex enterpr...           0   
10     asp net develop job opportun unit state new je...           0   
11     talent sourcer month fix term contract want bu...           0   
12     applic develop digit novitex enterpris solut f...           0   
13     instal grow event product compani provid stage...           0   
14     account execut sydney adthena ukâ lead competi...           0   
15     vp sale vault dragon jungl ventur lead singapo...           0   
16     hand qa leader honeybook weâ imagin event indu...           0   
17     southend sea traineeship na year old establish...           0   
18     visual design kettl independ digit agenc base ...           0   
19     process control engin dc plc m offic pa provid...           0   
20     market assist intellibright creat leverag ente...           0   
21     front end develop frustrat status quo like ima...           0   
22     engag manag upstreamâ mission revolutionis way...           0   
23     vice presid sale sponsorship businessfriend co...           0   
24     custom servic canari wharf base e commerc comp...           0   
25     h1b sponsor l1 l2 opt i28 technolog demonstr e...           0   
26     market exec work cubic seem like idea hell joi...           0   
27     haad dha licens doctor open uae medic recruit ...           0   
28     talent manag process manag provid full time pe...           0   
29     custom servic associ novitex enterpris solut f...           0   
...                                                  ...         ...   
17850  north west level dgv nvq assessor establish pr...           0   
17851  warehous oper materi handler red star yeast co...           0   
17852  gwt expert gwt expertw experienc rapid worldwi...           0   
17853  call center custom servic command care enough ...           0   
17854  talent manag process manag provid full time pe...           0   
17855  android engin shyp easiest way send anyth anyw...           0   
17856  web develop market intercom url_cb674f93cbcd21...           0   
17857  chief technolog offic cto fast grow hospit sta...           0   
17858  pharmaceut senior project manag consult keen36...           0   
17859  immedi requir io develop qatar greet vam syste...           0   
17860  sr sql server dba posit sr sql server dbaloc b...           0   
17861  junior embed engin tribe wearabl ventur back s...           0   
17862  clinic supervisor crossroad treatment center n...           0   
17863  implement support specialist design build sell...           0   
17864  next generat depth sens merl internship progra...           0   
17865  portfolio develop associ pari rest franc hire ...           0   
17866  post product editor stori relfrom foot look li...           0   
17867  admin assist seek administr work busi sale env...           0   
17868  cnc programm provid full time perman posit man...           0   
17869  sr technic lead lim job titl â sr technic lead...           0   
17870  senior financi analyst retail prosultingâ isâ ...           0   
17871  water truck driver valor servic provid workfor...           0   
17872  product manag flite deliv ad innov scale world...           0   
17873  recruit coordin respons facilit recruit hire p...           0   
17874  javascript de

In [21]:
data=data.drop(columns=['lemm', 'stemm'])

In [22]:
print(data[data['fraudulent']==1].shape)
data.dropna(inplace=True)
data[data['fraudulent']==1].shape
data

(866, 2)


text  fraudulent
0      market intern food52 creat groundbreak award w...           0
1      custom servic cloud video product second world...           0
2      commiss machineri assist cma valor servic prov...           0
3      account execut washington dc passion improv qu...           0
4      bill review manag spotsourc solut llc global h...           0
5      account clerk job overviewapex environment con...           0
6      head content f found theâ fonpit agâ rose inte...           0
7      lead guest servic specialist airenvyâ mission ...           0
8      hp bsm sme solutions3 woman own small busi who...           0
9      custom servic associ part time novitex enterpr...           0
10     asp net develop job opportun unit state new je...           0
11     talent sourcer month fix term contract want bu...           0
12     applic develop digit novitex enterpris solut f...           0
13     instal grow event product compani provid stage...           0
14     account execut sydney adthena ukâ lead competi...           0
15     vp sale vault dragon jungl ventur lead singapo...           0
16     hand qa leader honeybook weâ imagin event indu...           0
17     southend sea traineeship na year old establish...           0
18     visual design kettl independ digit agenc base ...           0
19     process control engin dc plc m offic pa provid...           0
20     market assist intellibright creat leverag ente...           0
21     front end develop frustrat status quo like ima...           0
22     engag manag upstreamâ mission revolutionis way...           0
23     vice presid sale sponsorship businessfriend co...           0
24     custom servic canari wharf base e commerc comp...           0
25     h1b sponsor l1 l2 opt i28 technolog demonstr e...           0
26     market exec work cubic seem like idea hell joi...           0
27     haad dha licens doctor open uae medic recruit ...           0
28     talent manag process manag provid full time pe...           0
29     custom servic associ novitex enterpris solut f...           0
...                                                  ...         ...
17850  north west level dgv nvq assessor establish pr...           0
17851  warehous oper materi handler red star yeast co...           0
17852  gwt expert gwt expertw experienc rapid worldwi...           0
17853  call center custom servic command care enough ...           0
17854  talent manag process manag provid full time pe...           0
17855  android engin shyp easiest way send anyth anyw...           0
17856  web develop market intercom url_cb674f93cbcd21...           0
17857  chief technolog offic cto fast grow hospit sta...           0
17858  pharmaceut senior project manag consult keen36...           0
17859  immedi requir io develop qatar greet vam syste...           0
17860  sr sql server dba posit sr sql server dbaloc b...           0
17861  junior embed engin tribe wearabl ventur back s...           0
17862  clinic supervisor crossroad treatment center n...           0
17863  implement support specialist design build sell...           0
17864  next generat depth sens merl internship progra...           0
17865  portfolio develop associ pari rest franc hire ...           0
17866  post product editor stori relfrom foot look li...           0
17867  admin assist seek administr work busi sale env...           0
17868  cnc programm provid full time perman posit man...           0
17869  sr technic lead lim job titl â sr technic lead...           0
17870  senior financi analyst retail prosultingâ isâ ...           0
17871  water truck driver valor servic provid workfor...           0
17872  product manag flite deliv ad innov scale world...           0
17873  recruit coordin respons facilit recruit hire p...           0
17874  javascript develop sr javascript develop exper...           0
17875  account director distribut vend look awesom ne...           0
17876  payrol account weblinc e commerc platform s

In [23]:
from sklearn.feature_extraction.text import TfidfVectorizer

# create the transform
vectorizer = TfidfVectorizer(stop_words='english')
X=vectorizer.fit_transform(data['text'])

In [32]:
def fitPredict(models_dict,X_train,y_train,X_test,y_test):
  acc={}
  spec={}
  sens={}
  model_save=None
  i=0
  #models_dict holds name and its function , like {'LR':LogsticRegression()}
  for name,model in models_dict.items():
    try:
      #fitting model on the train data
      model.fit(X_train,y_train)
      joblib.dump(model, name+'.pkl')
      #predicting model on X_test 
      model_save=model
      y_pred=model.predict(X_test)
      print('*'*100)
      print('Model:',name)
      #printing Confusion matrix an dAccuracy of each model
      print("Confusion Matrix: ")
      print(confusion_matrix(y_test, y_pred))
      print("Accuracy : ",accuracy_score(y_test,y_pred)*100)

      #mainitaing a dictionary of results of accuracy of each model to use it to plot in one systyem
      acc[name]=accuracy_score(y_test,y_pred)*100
      cm=confusion_matrix(y_test, y_pred)
      sens[name] = (cm[0,0]/(cm[0,0]+cm[0,1]))*100
      spec[name] = (cm[1,1]/(cm[1,0]+cm[1,1]))*100
      print("Classificarion Report : ")
      print(classification_report(y_test, y_pred))
      
    except ValueError:
      print(name,'failed')
  
  #converting dictionaries to dataframes to have them handy
  acc=pd.DataFrame.from_dict(acc,orient='index').reset_index()
  sens=pd.DataFrame.from_dict(sens,orient='index').reset_index()
  spec=pd.DataFrame.from_dict(spec,orient='index').reset_index()
  # print(accuracy)
  acc.columns=['model_name','accuracy']
  sens.columns=['model_name','sensitivity']
  spec.columns=['model_name','specificity']
  
  #plotting accuracy,sensitivity,specificity of each model
  return model_save

In [33]:
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer

from sklearn.externals import joblib 
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression,LinearRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from xgboost import XGBClassifier
from sklearn.cluster import KMeans
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, fbeta_score
from sklearn.naive_bayes import MultinomialNB

In [34]:
models_dict={}
models_dict['XGB']                      = XGBClassifier()

In [35]:
Y=data.loc[:,data.columns=='fraudulent']
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.15, random_state=1)
print(sum(y_train.fraudulent))

745


In [36]:
from imblearn.over_sampling import SMOTE
from imblearn.under_sampling import NearMiss
sm = SMOTE(random_state=12, sampling_strategy=0.5)
x_train_res, y_train_res = sm.fit_sample(X_train, y_train.fraudulent)
undersample = NearMiss(sampling_strategy=1)
x_train_final, y_train_final = undersample.fit_sample(x_train_res, y_train_res)

In [37]:
model_save=fitPredict(models_dict,X_train,y_train,X_test,y_test)

C:\Users\sdrsh\Anaconda3\lib\site-packages\sklearn\preprocessing\_label.py:235: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\sdrsh\Anaconda3\lib\site-packages\sklearn\preprocessing\_label.py:268: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


****************************************************************************************************
Model: XGB
Confusion Matrix: 
[[2559    2]
 [  43   78]]
Accuracy :  98.3221476510067
Classificarion Report : 
              precision    recall  f1-score   support

           0       0.98      1.00      0.99      2561
           1       0.97      0.64      0.78       121

    accuracy                           0.98      2682
   macro avg       0.98      0.82      0.88      2682
weighted avg       0.98      0.98      0.98      2682



In [38]:
model_save=fitPredict(models_dict,x_train_final,y_train_final,X_test,y_test)

****************************************************************************************************
Model: XGB
Confusion Matrix: 
[[2491   70]
 [  29   92]]
Accuracy :  96.30872483221476
Classificarion Report : 
              precision    recall  f1-score   support

           0       0.99      0.97      0.98      2561
           1       0.57      0.76      0.65       121

    accuracy                           0.96      2682
   macro avg       0.78      0.87      0.82      2682
weighted avg       0.97      0.96      0.97      2682



## Observation for the XGBoost algorithm using TF-IDF pipelining with Gensim vs NLTK and TF-IDF vectorization using under sampling and oversampling is


# Accuracy for XGBoost with TF-IDF Pipelining = 95.86


# Accuracy for XGboost with TF-IDF Vectorization without sampling = 98.322

# Accuracy for XGBoost with TF-IDF Vectorization with Undersampling majority class and oversampling minority class = 96.308